src/app/page.tsx

```diff
@@ -23,46 +24,73 @@ export default function Home() {
   };
 
   const handleTodoChange = async (id: number, state: ITodoItem) => {
+    const prevTodos = [...todos];
     if (state.title.trim() === "") {
       return;
     }
-    const newData = await updateData(id, state);
 
-    setTodos((todos) => todos.map((todo) => (todo.id === id ? newData : todo)));
+    setTodos((prevTodos) =>
+      prevTodos.map((todo) => (todo.id === id ? { ...todo, ...state } : todo))
+    );
+
+    try {
+      await updateData(id, state);
+    } catch (error) {
+      setTodos(prevTodos);
+    }
   };
 
   const handleAddTodo = async (todo: IAddTodoItemData) => {
     if (todo.title.trim() === "") {
+      setTitleError(true);
+      console.log("HIHIHIHI", titleError);
       return;
+    } else {
+      setTitleError(false);
+      console.log("HIHIHIHI", titleError);
     }
-
+    const prevTodos = [...todos];
     const newTodoData: IAddTodoItemData = {
       title: todo.title,
       completed: false,
     };
+    setTodos((prevTodos) => [
+      ...prevTodos,
+      { id: -1, createdAt: "", ...newTodoData },
+    ]);
 
     try {
-      const response = await addData(newTodoData);
-      setTodos([...todos, response]);
+      const addedTodo = await addData(newTodoData);
+
+      setTodos((prevTodos) =>
+        prevTodos.map((todo) => (todo.id === -1 ? { ...addedTodo } : todo))
+      );
     } catch (error) {
-      console.error("Failed to add new todo:", error);
+      setTodos(prevTodos);
+      setTitleError(true);
```

Comment:
If there is an error it could also be for a different reason than an empty value, right? Then I don't think you need to set the titleError here. You can add a generic error instead